lib/csv.rb

```diff
@@ -1252,74 +1253,123 @@ def shift
 
       next if @skip_lines and @skip_lines.match parse
 
-      parts =  parse.split(@col_sep_split_separator, -1)
-      if parts.empty?
+      scanner = StringScanner.new(parse)
+      if scanner.eos?
         if in_extended_col
-          csv[-1] << @col_sep   # will be replaced with a @row_sep after the parts.each loop
+          csv[-1] << @row_sep   # will be replaced with a @row_sep after the parts.each loop
         else
           csv << nil
         end
       end
 
       # This loop is the hot path of csv parsing. Some things may be non-dry
       # for a reason. Make sure to benchmark when refactoring.
-      parts.each do |part|
+      liberal_parsing_string = ""
+      until scanner.eos?
         if in_extended_col
-          # If we are continuing a previous column
-          if part.end_with?(@quote_char) && part.count(@quote_char) % 2 != 0
-            # extended column ends
-            csv.last << part[0..-2]
-            if csv.last.match?(@parsers[:stray_quote])
-              raise MalformedCSVError.new("Missing or stray quote",
-                                          lineno + 1)
+          if scanner.scan(@parsers[:quote])
+            if scanner.scan(@parsers[:quote])
+              csv.last << @quote_char
+              next
             end
-            csv.last.gsub!(@double_quote_char, @quote_char)
             in_extended_col = false
-          else
-            csv.last << part << @col_sep
-          end
-        elsif part.start_with?(@quote_char)
-          # If we are starting a new quoted column
-          if part.count(@quote_char) % 2 != 0
-            # start an extended column
-            csv << (part[1..-1] << @col_sep)
-            in_extended_col =  true
-          elsif part.end_with?(@quote_char)
-            # regular quoted column
-            csv << part[1..-2]
-            if csv.last.match?(@parsers[:stray_quote])
-              raise MalformedCSVError.new("Missing or stray quote",
-                                          lineno + 1)
+            if scanner.scan(@parsers[:col_sep])
+              csv << liberal_parsing_string if @liberal_parsing
+              if scanner.eos?
+                # e.g. %Q{a,"""\nb\n""",\nc}
+                csv << nil
+                break
+              end
+            elsif @liberal_parsing
+              csv << +"#{@quote_char}#{liberal_parsing_string}#{@quote_char}"
+              # e.g. '1,"\"2\"",3' #=> ["'1", "\"\\\"2\\\"\"", "3'"]
+              csv.last << scanner.scan(@parsers[:unquoted_value_liberal_parsing])
+              unless (scanner.eos? || scanner.scan(@parsers[:col_sep]))
```

Comment:
`next if scanner.eos? || scanner.scan(@parsers[:col_sep])` is preferred than `unless ||`.

Reply:
Certainly, I think so too. 104092e